File: Hashmap/p1_two_sum.py
class Solution(object):
    def twoSum(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: List[int]
        """
        
        hashmap={}
        for i, num in enumerate(nums):#separates to index, value
            leftover= target- num
            if leftover in hashmap:
                return[hashmap[leftover],i]

            hashmap[num]=i

    
'''

Explanation

the key is to create a hashmap of all values and their index to prevent double indexing
the leftover is returned first because it would exist previously


'''

File: Hashmap/p1_two_sum.cpp
class Solution {
public:
  vector<int> twoSum(vector<int> &nums, int target) {
    unordered_map<int, int> hashmap;
    for (int i = 0; i < nums.size(); i++) {
      int remain = target - nums[i];
      if (hashmap.count(remain)) {
        return {hashmap[remain], i};
      }
      hashmap[nums[i]] = i;
    }
    return {};
  }
};

// time complexity O(N)
// space complexity O(N)